### DISCUSSION TITLE
Help Bitcoin Public Key Why can't I convert the taproot public key through the private key?

### DISCUSSION THREAD

**Discussion by Xxxxxx-dot** [bitcoin-core/secp256k1#1602] [Q&A] [answered]
#include <stdio.h>
#include <stdlib.h>
#include <secp256k1.h>
#include <secp256k1_extrakeys.h>  

int main() {
    
    secp256k1_context *ctx = secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);

    
    unsigned char privkey[32] = {
        0x7d, 0x48, 0x1e, 0x43, 0x55, 0xd5, 0x1c, 0x82,
        0x4e, 0x8d, 0x02, 0xe6, 0x17, 0xd0, 0xe1, 0x88,
        0xba, 0x97, 0x68, 0x5e, 0x55, 0xf8, 0x65, 0x57,
        0xd1, 0xd8, 0x3d, 0x64, 0xd1, 0xc8, 0x95, 0x4b
    };
	
    
    if (!secp256k1_ec_seckey_verify(ctx, privkey)) {
        printf("Invalid private key\n");
        secp256k1_context_destroy(ctx);
        return 1;
    }

    
    secp256k1_pubkey pubkey;

    
    if (!secp256k1_ec_pubkey_create(ctx, &pubkey, privkey)) {
        printf("Failed to create public key\n");
        secp256k1_context_destroy(ctx);
        return 1;
    }

    
    secp256k1_xonly_pubkey xonly_pubkey;
    if (!secp256k1_xonly_pubkey_from_pubkey(ctx, &xonly_pubkey, NULL, &pubkey)) {
        printf("Failed to convert to x-only public key\n");
        secp256k1_context_destroy(ctx);
        return 1;
    }

    unsigned char output[32];
    if (!secp256k1_xonly_pubkey_serialize(ctx, output, &xonly_pubkey)) {
        printf("Failed to serialize x-only public key\n");
        secp256k1_context_destroy(ctx);
        return 1;
    }

    
    printf("Taproot public key (x-only): ");
    for (size_t i = 0; i < 32; i++) {
        printf("%02x", output[i]);
    }
    printf("\n");

    
    secp256k1_context_destroy(ctx);

    return 0;
}

I need t2pr public key

I need an example, can anyone help me?

**Comment by Xxxxxx-dot** [comment]
Such a conversion is wrong taproot public key

**Comment by jonasnick** [comment]
What's the problem you're having with that code?

**Reply by Xxxxxx-dot** [reply]
Hello I am trying to convert p2tr private key to p2tr public key. But this result, the output of the above code is wrong, after comparison with C# NBitcoin. The complete public keys are consistent, but the taproot public keys are inconsistent. I don’t know what else needs to be done.

**Reply by Xxxxxx-dot** [reply]
The correct result should be  

Get Taproot FullPubKey: 9161099f0ea8b9bb631e318cb9ab0b1811319e1d2e7ac1aa68fcda9c5c85c4a3

**Reply by jonasnick** [reply]
The code you posted above looks correct to me.

**Reply by Xxxxxx-dot** [reply]
Taproot public key (x-only): 8bd5ac9306da862cc9cd1b95ae4871dd27054e49774f6319221ab8dc262d9fd7  If this result is converted into an address, it is a wrong address. This result is the result of running the above code.

Can you take a look at why it's different? I'm sure the NBitcoin result of the C# library is correct.

**Reply by jonasnick** [reply]
All I can say is that your code correctly uses libsecp256k1 to convert the given secret key into a a serialized xonly public key.

**Reply by Xxxxxx-dot** [reply]
Is the serialized xonly public key equal to taproot's public key? I'm willing to spend a little money to get help.

**Reply by sipa** [reply]
Taproot outputs typically have an *internal* public key which differs from the *output* public key. In outputs that wish to be spendable through both a key path and a script path, this tweaking must follow the exact consensus rules specified in BIP341 (or the script path spending will not be available). For outputs that are only intended to be spent through the key path, this tweaking is optional (but there is a still a common, standard, recommended, way of doing so described in BIP341).

Your code is not doing any tweaking. I have not verified that the public key you expect is perhaps the standard key-path-only tweaking for the private/public key you give, but I suspect that may be it.

**Reply by Xxxxxx-dot** [reply]
Hello, can you give an example? Thank you so much.

**Reply by Xxxxxx-dot** [reply]
secp256k1_context* ctx = secp256k1_context_create(SECP256K1_CONTEXT_SIGN);


unsigned char privkey[32] = {
    0x7d, 0x48, 0x1e, 0x43, 0x55, 0xd5, 0x1c, 0x82,
        0x4e, 0x8d, 0x02, 0xe6, 0x17, 0xd0, 0xe1, 0x88,
        0xba, 0x97, 0x68, 0x5e, 0x55, 0xf8, 0x65, 0x57,
        0xd1, 0xd8, 0x3d, 0x64, 0xd1, 0xc8, 0x95, 0x4b
};


secp256k1_keypair keypair;


if (!secp256k1_keypair_create(ctx, &keypair, privkey)) {
    printf("NO-0\n");
    return 1;
}


secp256k1_xonly_pubkey xonly_pubkey;
int pk_parity;
if (!secp256k1_keypair_xonly_pub(ctx, &xonly_pubkey, &pk_parity, &keypair)) {
    printf("NO-1\n");
    return 1;
}


unsigned char tweak[32] = { 0 }; //  tweak 0 ???


secp256k1_pubkey tweaked_pubkey;
if (!secp256k1_xonly_pubkey_tweak_add(ctx, &tweaked_pubkey, &xonly_pubkey, tweak)) {
    printf("NO-2\n");
    return 1;
}


unsigned char tweaked_pubkey33[33];
size_t output_len = sizeof(tweaked_pubkey33);
secp256k1_ec_pubkey_serialize(ctx, tweaked_pubkey33, &output_len, &tweaked_pubkey, SECP256K1_EC_COMPRESSED);

  printf("BIP341 Taproot : ");
for (int i = 0; i < output_len; i++) {
    printf("%02x", tweaked_pubkey33[i]);
}
  printf("\n");

  
 secp256k1_context_destroy(ctx);

 return 0;

**Reply by Xxxxxx-dot** [reply]
I adjusted it like this and the result is still wrong.

**Comment by CASABECI** [comment]
> #include <stdio.h> #include <stdlib.h> #include <secp256k1.h> #include <secp256k1_extrakeys.h>
> 
> int main() {
> 
> ```
> secp256k1_context *ctx = secp256k1_context_create(SECP256K1_CONTEXT_SIGN | SECP256K1_CONTEXT_VERIFY);
> 
> 
> unsigned char privkey[32] = {
>     0x7d, 0x48, 0x1e, 0x43, 0x55, 0xd5, 0x1c, 0x82,
>     0x4e, 0x8d, 0x02, 0xe6, 0x17, 0xd0, 0xe1, 0x88,
>     0xba, 0x97, 0x68, 0x5e, 0x55, 0xf8, 0x65, 0x57,
>     0xd1, 0xd8, 0x3d, 0x64, 0xd1, 0xc8, 0x95, 0x4b
> };
> 
> 
> if (!secp256k1_ec_seckey_verify(ctx, privkey)) {
>     printf("Invalid private key\n");
>     secp256k1_context_destroy(ctx);
>     return 1;
> }
> 
> 
> secp256k1_pubkey pubkey;
> 
> 
> if (!secp256k1_ec_pubkey_create(ctx, &pubkey, privkey)) {
>     printf("Failed to create public key\n");
>     secp256k1_context_destroy(ctx);
>     return 1;
> }
> 
> 
> secp256k1_xonly_pubkey xonly_pubkey;
> if (!secp256k1_xonly_pubkey_from_pubkey(ctx, &xonly_pubkey, NULL, &pubkey)) {
>     printf("Failed to convert to x-only public key\n");
>     secp256k1_context_destroy(ctx);
>     return 1;
> }
> 
> unsigned char output[32];
> if (!secp256k1_xonly_pubkey_serialize(ctx, output, &xonly_pubkey)) {
>     printf("Failed to serialize x-only public key\n");
>     secp256k1_context_destroy(ctx);
>     return 1;
> }
> 
> 
> printf("Taproot public key (x-only): ");
> for (size_t i = 0; i < 32; i++) {
>     printf("%02x", output[i]);
> }
> printf("\n");
> 
> 
> secp256k1_context_destroy(ctx);
> 
> return 0;
> ```
> 
> }
> 
> I need t2pr public key
> 
> I need an example, can anyone help me?

**Reply by Xxxxxx-dot** [reply]
Thank you very much, but the result is still incorrect  The two results do not match
C# NBitcoin 》》FullPubKey: 9161099f0ea8b9bb631e318cb9ab0b1811319e1d2e7ac1aa68fcda9c5c85c4a3

Taproot public key (x-only): 8bd5ac9306da862cc9cd1b95ae4871dd27054e49774f6319221ab8dc262d9fd7

**Comment by NicolasDorier** [comment, marked as answer]
In NBitcoin

```csharp
var privBytes = new byte[] { 0x7d, 0x48, 0x1e, 0x43, 0x55, 0xd5, 0x1c, 0x82,
0x4e, 0x8d, 0x02, 0xe6, 0x17, 0xd0, 0xe1, 0x88,
0xba, 0x97, 0x68, 0x5e, 0x55, 0xf8, 0x65, 0x57,
0xd1, 0xd8, 0x3d, 0x64, 0xd1, 0xc8, 0x95, 0x4b};
var key = new Key(privBytes);
Console.WriteLine("33 bytes pubkey: " + key.PubKey.ToHex());
var full = key.PubKey.GetTaprootFullPubKey();
Console.WriteLine("32 bytes output key: " + full.OutputKey.ToString());
Console.WriteLine("32 bytes internal key: " + full.InternalKey.ToString());
```

Result:

```
33 bytes pubkey: 028bd5ac9306da862cc9cd1b95ae4871dd27054e49774f6319221ab8dc262d9fd7
32 bytes output key: 9161099f0ea8b9bb631e318cb9ab0b1811319e1d2e7ac1aa68fcda9c5c85c4a3
32 bytes internal key: 8bd5ac9306da862cc9cd1b95ae4871dd27054e49774f6319221ab8dc262d9fd7
```

`8bd...` is the `internal key` and `91...` is the `output key`. If you don't understand the difference, please refer to the BIPs concerning taproot.

**Reply by Xxxxxx-dot** [reply]
Thank you very much for your help sir

**Reply by Xxxxxx-dot** [reply]
The problem now is that I cannot achieve the   output key

If possible can you give me an example?
